Add TTOI functions

Lili-Zheng-stat opened the pull request:
I just created a new branch in my repository that is updated with the current version of TensorLy. The updated tensor train functions (`tt_TTOI.py` in tensorly/tensorly/contrib/decomposition, and `test_tt_TTOI.p`y in tensorly/tensorly/contrib/decomposition/tests) are also added and pass the tests with `backend = numpy` or `pytorch`.

JeanKossaifi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,188 @@
+        and len(rank) == len(tl.shape(data_tensor))+1
+    n_iter : int
+        half the number of iterations
+    trajectory : bool, optional, default is False
```
Is there a practical case where we'd actually want this? It would be quite memory heavy.

JeanKossaifi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,188 @@
+    for n in range(n_iter):
+        # first perform forward update
+        # U_arr will be a list including estimated left singular spaces at the current iteration
+        U_arr = list()
```
A general comment on variable names: we want to use explicit names, that make it clear to the reader what the variable does. e.g. _arr is unnecessary. Could we name it e.g. `unfolding_eigenvals` or something like that?

Lili-Zheng-stat replied:
Right... I have changed U_arr to left_singular_vectors and V_arr to right_singular_vectors

JeanKossaifi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,188 @@
+        If direction == "right" and `multiplier_list` contains d-N tensors, then `tensor_prod` is of dimension r_0*p_1*...*p_N*r_N
+    """ 
+
+    if direction == "left":
```
Not sure we need a function just for that but if we do decide to keep it, perhaps we can just have left_to_right=True by default and just set it to False to to right to left?

Lili-Zheng-stat replied:
Sounds good! I have changed this parameter to left_to_right with the default being True.

JeanKossaifi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,188 @@
+        error_list = list()
+        
+    # perform TTOI for n_iter iterations
+    for n in range(n_iter):
```
```suggestion
    for iteration in range(n_iter):
```

JeanKossaifi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,188 @@
+        U_arr.append(tl.reshape(U_tmp,(rank[0],shape[0],rank[1])))
+
+        # estimate the 2nd to (d-1)th left singular spaces
+        for k in range(n_dim-2):
```
```suggestion
        for mode in range(n_dim-2):
```
same here, good to have explicit names

JeanKossaifi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,188 @@
+            U_arr.append(tl.reshape(U_tmp,(rank[k+1],shape[k+1],rank[k+2])))
+
+        # forward update is done; output the final residual
+        R_tilde_arr.append(sequential_prod(R_tmp_l,[U_arr[n_dim-2]],"left"))
```
Same for R_tilde_arr, etc good to try to find informative names.

Lili-Zheng-stat replied:
Sounds good! Changed to left_residuals.

JeanKossaifi reviewed (commented):
Thanks @Lili-Zheng-stat, it's much clearer! I left some mostly stylistic comments in the code. Would be great to apply pep8 formatting too.
Looking to have TTOI in TensorLy!

JeanKossaifi commented:
Thanks @Lili-Zheng-stat!

The issue in the tests comes from the fact that we changed the SVD interface in #429, the best would be to use `tensorly.tenalg.svd_interface`. @aarmey perhaps we should deprecate tl.partial_svd or at least raise an error if users try to use it, to let them know they should switch to the new interface?

aarmey commented:
> Thanks @Lili-Zheng-stat!
> 
> The issue in the tests comes from the fact that we changed the SVD interface in #429, the best would be to use `tensorly.tenalg.svd_interface`. @aarmey perhaps we should deprecate tl.partial_svd or at least raise an error if users try to use it, to let them know they should switch to the new interface?

Good idea. I'll add a deprecation before we think about releasing this.

Lili-Zheng-stat commented:
> Thanks @Lili-Zheng-stat!
> 
> The issue in the tests comes from the fact that we changed the SVD interface in #429, the best would be to use `tensorly.tenalg.svd_interface`. @aarmey perhaps we should deprecate tl.partial_svd or at least raise an error if users try to use it, to let them know they should switch to the new interface?

Thanks for checking the errors! I updated the code so that it calls the new svd_interface now.

caglayantuna commented:
You can format your files with Black. Then, it will pass the test. To use Black;

```
pip install black
black filename.py
```

JeanKossaifi commented:
Thanks @Lili-Zheng-stat -- sorry for the delay, I'll try to make another review asap. Would be great if someone else could also have a look, maybe @merajhashemi since you're familiar with TT too?

cohenjer commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,255 @@
+
+    if left_to_right == True:
+        for i in range(len(multiplier_list)):
+            tensor_prod = tl.tensordot(
```
Just wondering if instead of tensordot, it would be possible to use the n_mode_product which, in my experience of Tensorly, is often faster. Tensordot calls the backend's tensordot directly with n_mode_product might benefit from any improved implementaton of tensor contractions we might add to Tensorly in the future.

Lili-Zheng-stat replied:
It seems that the n_mode_product function is for tensor-matrix product or tensor-vector product; is that correct? I am using tensordot here for sequential tensor product, and hence would prefer not to change this function if possible....

cohenjer replied:
You are right, ignore my comment!

cohenjer reviewed (commented):
Hi @Lili-Zheng-stat, the code looks nice and the method is cool :)
I left a few comments. Maybe it could be nice to see some example of the code in action, possibly added to the Tensorly examples? But that can be done later in a subsequent PR.

JeanKossaifi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,255 @@
+        # perform backward update
+        # initialize right_singular_vectors: right_singular_vectors will be a list of estimated right singular spaces at the current or previous iteration
+        right_singular_vectors = list()
+        V_tmp = tl.transpose(
```
Same here, I would do _, _, V_tmp = ...

Lili-Zheng-stat replied:
Changed this, too!

Lili-Zheng-stat commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,255 @@
+def tensor_train_OI(
+    data_tensor, rank, n_iter=1, trajectory=False, return_errors=True, **context
+):
+    """Perform tensor-train orthogonal iteration (TTOI) for tensor train decomposition
```
Changed this!

Lili-Zheng-stat commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,255 @@
+    * one tensor (otherwise): output of the last iteration, one full tensor and list of tensor factors 
+    full_tensor : list of n_iter tensors or one tensor
+    * n_iter tensors (if `trajectory` is True) : each list contains the output of each iteration, one full tensor and list of tensor factoros
+    * one tensor (otherwise): output of the last iteration, one full tensor and list of tensor factors 
```
Thanks for pointing this out!! I have made some corrections here

cohenjer replied:
still small typo (factoros) :)

Lili-Zheng-stat commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,255 @@
+    for iteration in range(n_iter):
+        # first perform forward update
+        # left_singular_vectors will be a list including estimated left singular spaces at the current iteration
+        left_singular_vectors = list()
```
Thanks for the suggestion! However, our later updates make use of many intermediate product obtained during the TT-SVD algorithm, and it would need a lot of changes in our code structure if we want to use the default TT-SVD function.... If possible, I would prefer not changing this now...

merajhashemi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,260 @@
+            factors_list_tmp = []
+            for mode in range(n_dim - 1):
+                factors_list_tmp.append(
+                    tl.tensor(left_singular_vectors[mode], **context)
```
There is no need to wrap this inside tl.tensor.

merajhashemi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,260 @@
+                    tl.tensor(left_singular_vectors[mode], **context)
+                )
+            factors_list_tmp.append(tl.tensor(left_residuals[n_dim - 2], **context))
+            full_tensor_tmp = tl.tensor(tt_to_tensor(factors_list_tmp), **context)
```
tl.tensor can be removed

merajhashemi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,260 @@
+                            **context
+                            )
+                        )
+                full_tensor_tmp = tl.tensor(tt_to_tensor(factors_list_tmp), **context)
```
Same. No need for tl.tensor

merajhashemi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,260 @@
+                factors_list_tmp.append(tl.tensor(Residual_right, **context))
+                for mode in range(n_dim - 1):
+                    factors_list_tmp.append(
+                        tl.tensor(
```
tl.tensor can be removed

merajhashemi reviewed (commented):
I left a few comments. There are some`tl.tensor`s which I think are not necessary. Removing them could improve code readability.

merajhashemi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,260 @@
+                factors_list_tmp.append(
+                    tl.tensor(left_singular_vectors[mode], **context)
+                )
+            factors_list_tmp.append(tl.tensor(left_residuals[n_dim - 2], **context))
```
tl.tensor can be removed

merajhashemi commented on tensorly/contrib/decomposition/tt_TTOI.py:
```diff
@@ -0,0 +1,260 @@
+                )
+            if trajectory or return_errors or iteration == np.ceil(n_iter/2) - 1:
+                factors_list_tmp = []
+                factors_list_tmp.append(tl.tensor(Residual_right, **context))
```
Same

Lili-Zheng-stat replied:
> I left a few comments. There are some`tl.tensor`s which I think are not necessary. Removing them could improve code readability.

Thanks for the suggestions. I have removed them.

JeanKossaifi commented:
It looks good to me, thanks for all the change @Lili-Zheng-stat -- it looks good to me. Can you just skip the test for MXNet (and maybe add a small test on 3rd or 4th order tensor). MXNet currently doesn't support >6th order tensors. 

Can you also add the new function to the [API](https://github.com/tensorly/tensorly/blob/main/doc/modules/api.rst)?

I think we can merge it afterwards!

JeanKossaifi commented:
@Lili-Zheng-stat apologies, the tests are failing due to an unrelated issue, could you merge main, that should fix the issue! Thanks!

Lili-Zheng-stat commented:
> Sorry that I just saw this! It seems that you have merged main already? Let me know if there is anything else I need to do at this point. Thanks!!

JeanKossaifi commented:
Yes, I realized you enabled edit from maintainers so I went ahead and merged / fixed the issue, looks like tests are passing now, merging.

Thanks a lot for the PR @Lili-Zheng-stat, this is a great addition! It would be great to add an example in a separate PR.
And thanks @cohenjer @merajhashemi @aarmey @caglayantuna for the reviews!

Lili-Zheng-stat commented:
> 

Awesome!! Thanks to everyone @JeanKossaifi @cohenjer @merajhashemi @aarmey @caglayantuna for the review and all the suggestions!
